clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/operations/ConditionalReplaceOperation.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.clustered.client.internal.store.operations;
+
+import org.ehcache.clustered.client.internal.store.operations.codecs.CodecException;
+import org.ehcache.spi.serialization.Serializer;
+
+import java.nio.ByteBuffer;
+
+import static org.ehcache.clustered.client.internal.store.operations.OperationCode.REPLACE_CONDITIONAL;
+
+public class ConditionalReplaceOperation<K, V> extends BaseOperation<K, V> {
+
+  private final V oldValue;
+
+  public ConditionalReplaceOperation(final K key, final V oldValue, final V value) {
+    super(key, value);
+    if(oldValue == null) {
+      throw new NullPointerException("Old value can not be null");
+    }
+    this.oldValue = oldValue;
+  }
+
+  ConditionalReplaceOperation(final ByteBuffer buffer, final Serializer<K> keySerializer, final Serializer<V> valueSerializer) {
+    OperationCode opCode = OperationCode.valueOf(buffer.get());
```

Comment:
The fact that you have to write all this here, plus the default no arg constructor on `BaseOperation` plus the protected fields are an indication that something is wrong in the abstraction you defined.
I think a lot of these problems would go away if the BaseOperation did not deal at all with values and that is left to subclasses. The the fact that you have one or two values can be properly encapsulated.